### PR TITLE
Cards winrates

### PR DESCRIPTION
**Pull request by Manuel-777**
Cards Winrate data display on Decks view as requested on Discord;

image

I could not find any relevant issues regarding this, altrough I remember it being mentioned and requested quite a couple times.

### CODE REVIEW

**Comment by riQQ** [comment]
Is Winrate the played winrate?

**Comment by Manuel-777** [comment]
> Is Winrate the played winrate?

Yes, its only added to the calculation if the card in question was cast (except on the "sideboarded out" part)

File: src/shared/colors.ts [hunk 1]
```diff
@@ -131,6 +143,19 @@ class Colors {
     return this;
   }
 
+  /**
+   * Merges a "bitshift" integer into this color.
+   */
+  addFromBits(color: number) {
```
Comment (Thaoden) [inline src/shared/colors.ts:149]: Shouldn't this be `getBits(color)`?

File: src/shared/colors.ts [hunk 2]
```diff
@@ -26,16 +26,28 @@ class Colors {
...
     if (this.g !== 0) _arr.push(GREEN);
-
     return _arr;
   }
 
+  /**
+   * Returns an Integer that identifies this color as a "bitshift" sum
+   */
+  getBits(): number {
+    let sum = 0;
```
Comment (Thaoden) [inline src/shared/colors.ts:42]: nit: As this is now extracted into its own method, maybe rename `sum` to `bits` to make its intent more clear.

File: src/window_main/aggregator.ts [hunk 3]
```diff
@@ -486,4 +523,88 @@ export default class Aggregator {
...
+          // For each card cast
+          game.cardsCast.forEach(cardCast => {
+            const { grpId, player, turn } = cardCast;
+            // Only if we casted it
+            if (player == match.player.seat) {
+              // define
+              if (!winrates[grpId]) winrates[grpId] = newCardWinrate(grpId);
+              // Only once per card cast!
+              if (!gameCards.includes(grpId)) {
+                winrates[grpId].wins += game.win ? 1 : 0;
```
Comment (Thaoden) [inline src/window_main/aggregator.ts:549]: This ternary appears multiple times identically in this block. Maybe introduce some temp variables and set them once at the beginning of the block.

File: src/window_main/aggregator.ts [hunk 4]
```diff
@@ -486,4 +523,88 @@ export default class Aggregator {
...
+                    };
+                  }
+                  winrates[grpId].colors[bits].wins += game.win ? 1 : 0;
+                  winrates[grpId].colors[bits].losses += game.win ? 0 : 1;
+                  gameCards.push(grpId);
+                }
+              }
+              // Do this for every card cast in the game
+              winrates[grpId].turnsUsed.push(
+                turn % 2 ? (turn + 1) / 2 : turn / 2
```
Comment (Thaoden) [inline src/window_main/aggregator.ts:569]:
```suggestion
Math.ceil(turn / 2);
```

File: src/window_main/components/deck-view/CardsWinrateView.tsx [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+  winrates: Record<number, CardWinrateData>,
+  cardObj: DbCardData,
+  quantity: number,
+  index: number
+): JSX.Element {
+  const wr = winrates[cardObj.id];
+  const winrate =
+    wr.wins + wr.losses == 0
+      ? -1
+      : Math.round((100 / (wr.wins + wr.losses)) * wr.wins);
```
Comment (Thaoden) [inline src/window_main/components/deck-view/CardsWinrateView.tsx:22]: Maybe extract out a method for this; this calculation is made several times (with different params).

**Comment by riQQ** [comment]
> 
> 
> > Is Winrate the played winrate?
> 
> Yes, its only added to the calculation if the card in question was cast (except on the "sideboarded out" part)

Maybe it would be good to rename it to Cast WR then.

File: src/window_main/aggregator.ts (hunk 4 above)
Comment (Manuel-777) [reply src/window_main/aggregator.ts:569]: I was definitely overthinking this one lol

File: src/shared/colors.ts (hunk 1 above)
Comment (Manuel-777) [reply src/shared/colors.ts:149]: Not, since this would be the inverse method of getBits

Comment (Thaoden) [reply src/shared/colors.ts:149]: So `color` is already in bitmap form?

File: src/shared/colors.ts [hunk 6]
```diff
@@ -40,11 +40,11 @@ class Colors {
    */
   getBits(): number {
     let bits = 0;
-    if (this.w !== 0) bits = bits | 2;
-    if (this.u !== 0) bits = bits | 4;
-    if (this.b !== 0) bits = bits | 8;
-    if (this.r !== 0) bits = bits | 16;
-    if (this.g !== 0) bits = bits | 32;
+    if (this.w !== 0) bits = bits | 1;
```
Comment (Thaoden) [inline src/shared/colors.ts:43]:
```suggestion
if (this.w !== 0) bits |= 1;
```
etc.
Though I think `1` was undefined, not white?

Comment (Thaoden) [reply src/shared/colors.ts:43]: Also, you could promote the `colorFlags` array from line 148 to a class local array and reuse it here.

File: src/shared/colors.ts (hunk 1 above)
Comment (Manuel-777) [reply src/shared/colors.ts:149]: Yes, `addFromBits` takes a bitmap and includes it into this `Color`

Comment (Thaoden) [reply src/shared/colors.ts:157]: Maybe rename `color` (the value passed in) to `colorBitmap` or sth, to make its intent more clear. Only a nit though...

**Pull request merged**